dataplane/snappi/sai_dataplane.py

```diff
@@ -0,0 +1,334 @@
+import logging
+import time
+
+import dpkt
+from saichallenger.common.sai_dataplane import SaiDataplane
+from snappi import snappi
+
+BASE_TENGINE_PORT = 5555
+
+class SaiDataplaneImpl(SaiDataplane):
+
+    def __init__(self, exec_params):
+        self.alias = exec_params['alias']
+        self.mode = exec_params['mode']
+        super().__init__(exec_params)
+        self.flows = []
+
+    def init(self):
+        # Configure a new API instance where the location points to controller
+        # Ixia-C:       location = "https://<tgen-ip>:<port>"
+        # IxNetwork:    location = "https://<tgen-ip>:<port>", ext="ixnetwork"
+        # TRex:         location =         "<tgen-ip>:<port>", ext="trex"
+        if self.mode == 'ixnet':
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False, ext='ixnetwork')
+        elif self.mode == 'trex':
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False, ext='trex')
+        else:
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False)
+        logging.info("%s Starting connection to controller... " % time.strftime("%s"))
+
+        # Create an empty configuration to be pushed to controller
+        self.configuration = self.api.config()
+
+        # Configure two ports where the location points to the port location:
+        # Ixia-C:       port.location = "localhost:5555"
+        # IxNetwork:    port.location = "<chassisip>;card;port"
+        # TRex:         port.location = "localhost:5555"
+        for pid, port in enumerate(self.exec_params['port_groups']):
+            location = port.get('location', f"localhost:{BASE_TENGINE_PORT+pid}")
+            self.configuration.ports.port(name=port[port['init']], location=location)
+
+        cap = self.configuration.captures.capture(name="c1")[-1]
+        cap.port_names = [p.name for p in self.configuration.ports]
+        cap.format = cap.PCAP
+
+        self.dataplane = self
+
+    def remove(self):
+        pass
+
+    def setUp(self):
+        super().setUp()
+        self.set_config()
+        self.start_capture()
+
+    def tearDown(self):
+        super().tearDown()
+        self.stop_capture()
+        self.configuration.flows.clear()
+        self.flows.clear()
+
+    @staticmethod
+    def api_results_ok(results):
+        if hasattr(results, 'warnings'):
+            return True
+        else:
+            print(f'Results={results}')
+            return False
+
+    @staticmethod
+    def check_warnings(response):
+        if response.warnings:
+            print("Warning: %s" % str(response.warnings))
+
+    @staticmethod
+    def get_capture_port_names(cfg):
+        """
+        Returns name of ports for which capture is enabled.
+        """
+        names = []
+        for cap in cfg.captures:
+            if cap._properties.get("port_names"):
```

Comment:
would suggest not using private members here; should use `if cap.port_names` instead.

Reply:
fixed